### PR TITLE
ML-454 GNN Version 2.0

### PR DESCRIPTION
**Pull request by RogerDev**
Add Multi-model support
Add Keras Functional API for complex models
Add Multiple input and output support for models
Add Tensorflow 2.x as well as 1.x support
Add GPU support
Add trainToLoss
Add Learning Rate reduction
Add Batch Size Reduction
Optimize MakeTensor()
Fix zero tensor issue
Fix performance problem with tensor weights synchronization
Update and improve documentation

Signed-off-by: Roger Dev <Roger.Dev@lexisnexisrisk.com>

### CODE REVIEW

File: GNNI.ecl [hunk 1]
```diff
@@ -304,75 +449,124 @@ EXPORT GNNI := MODULE
...
+    maxInputWi := MAX(x, wi);
+    // Change the wi's for outputs (y) so that they are after the input wi's
+    y1 := PROJECT(y, TRANSFORM(RECORDOF(LEFT), SELF.wi := LEFT.wi + maxInputWi, SELF := LEFT), LOCAL);
+    aligned := Tensor.R4.AlignTensors(x + y1);
+    // Now change the Y's wi back to the original numbers
+    xAl := aligned(wi <= maxInputWi);
+    yAl := PROJECT(aligned(wi > maxInputWi), TRANSFORM(RECORDOF(LEFT), SELF.wi := LEFT.wi - maxInputWi, SELF := LEFT), LOCAL);
     totalRecords := Tensor.R4.GetRecordCount(yAl);
-    batchesPerEpoch := ROUNDUP(totalRecords / nNodes / batchSize);
+    //batchesPerEpoch := ROUNDUP(totalRecords / nNodes / batchSize);
```
Comment (ghalliday) [inline GNNI.ecl]: ?Should this be deleted?

File: GNNI.ecl [hunk 2]
```diff
@@ -304,75 +449,124 @@ EXPORT GNNI := MODULE
...
         wtChanges := DISTRIBUTE(wtChanges0, wi + sliceId);
         // Sum up the original weights (de-replicated) and all changes for each wi and slice
         newWts := rollUpdates(wts2((wi + sliceId) % nNodes = nodeId), wtChanges);
         // Note: newWts have been replicated to all nodes by rollUpdates.
-        // We use epochNum + batchNum to generate a unique model token for
-        // use with GetLoss.  This ensures proper sequencing of the
-        // operation.
-        batchLoss := IF(EXISTS(newWts), GetLoss(model + epochNum + batchNum), 1.0);
-        logProgress2 := Syslog.addWorkunitInformation('Training Status (2): Epoch = ' + epochNum + ', Batch = ' + batchNum + ', Loss = ' + batchLoss);
+        // We use obscure to prevent the ECL compiler from treating GetLoss as a
```
Comment (ghalliday) [inline GNNI.ecl]: It doesn't look like obscure is used any more

File: GNNI.ecl [hunk 3]
```diff
@@ -304,75 +449,124 @@ EXPORT GNNI := MODULE
...
+        logProgress2 := Syslog.addWorkunitInformation('Training Status (2): ModelId = ' +
+                kModelId + ', Epoch = ' + epochNum + ', Batch = ' + batchNum + ', Loss = ' + batchLoss);
         RETURN newWts;
       END;
-      epochWts := LOOP(wts1, batchesPerEpoch, doBatch(ROWS(LEFT), COUNTER));
-      epochLoss := IF(EXISTS(epochWts), GetLoss(model + epochNum), 1.0);
-      logProgress := Syslog.addWorkunitInformation('Training Status: Epoch = ' + epochNum + ', Loss = ' + epochLoss);
+      epochWts0 := LOOP(wts1, batchesPerEpoch, doBatch(ROWS(LEFT), COUNTER));
+      epochLoss := IF(EXISTS(epochWts0), GetLoss(model + (batchesPerEpoch * (epochNum-1))), 1.0);
+      //epochLoss := IF(EXISTS(epochWts), GetLoss(obscure(model)), 1.0);
```
Comment (ghalliday) [inline GNNI.ecl]: can this be deleted?

File: Internal/TensExtract.ecl [hunk 4]
```diff
@@ -8,145 +8,177 @@ nNodes := Thorlib.nodes();
 
 t_Tensor := Tensor.R4.t_Tensor;
 
-MAX_SLICE := Tensor.MAX_SLICE;
+//MAX_SLICE := Tensor.MAX_SLICE;
+MAX_SLICE := POWER(2, 24);
```
Comment (ghalliday) [inline Internal/TensExtract.ecl:11]: One of these should probably be removed.

File: Tensor.ecl [hunk 5]
```diff
@@ -586,25 +586,30 @@ EXPORT Tensor
       slicesPerNode := nSlices / nNodes;
       relNode := (sliceId - 1) div slicesPerNode;
       // Offset per wi so that differnt work items use staggered node numbering
-      nodeId := (wi - 1) + relNode % nNodes;
+      //nodeId := (wi - 1) + relNode % nNodes;
+      nodeId := relNode % nNodes; // Temporarily disable spreading by wi
```
Comment (ghalliday) [inline Tensor.ecl]: should one of these be removed/commented?

File: Test/ClassicTest.ecl [hunk 6]
```diff
@@ -85,8 +90,8 @@ trainY0 := NORMALIZE(train, 1, TRANSFORM(TensData,
 trainX := Tensor.R4.MakeTensor([0, featureCount], trainX0);
 trainY:= Tensor.R4.MakeTensor([0, 1], trainY0);
 
-OUTPUT(trainX, NAMED('X1'));
-OUTPUT(trainY, NAMED('y1'));
+//OUTPUT(trainX[.. 10000], NAMED('X1'));
+//OUTPUT(trainY[.. 10000], NAMED('y1'));
```
Comment (ghalliday) [inline Test/ClassicTest.ecl]: Should these remain commented out?

File: Test/FuncModelTest.ecl [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+// Break the training and test data into X (independent) and Y (dependent) data sets.
+// Format as NumericField data.
+trainX0C := NORMALIZE(trainC, featureCount, TRANSFORM(TensData,
+                            SELF.indexes := [LEFT.id, COUNTER],
+                            SELF.value := LEFT.x[COUNTER]));
+trainY0C := NORMALIZE(trainC, 1, TRANSFORM(TensData,
+                            SELF.indexes := [LEFT.id, COUNTER],
+                            SELF.value := LEFT.Y));
+// We need to one-hot enncode the Y value since we are doing classification using
```
Comment (ghalliday) [inline Test/FuncModelTest.ecl]: typo? enncode?

**Review by ghalliday — COMMENTED** [review]
@RogerDev a few minor comments.

**Comment by RogerDev** [comment]
@ghalliday Made changes based on your code review.  Please review and merge.

**Comment by ghalliday** [comment]
@RogerDev thanks.  Please squash and I will merge.

**Comment by RogerDev** [comment]
Squashed and ready to merge
